proposals/0064-transaction-receipt.md

```diff
@@ -0,0 +1,257 @@
+  / |     / \
+ /  |    /   \
+L0  L1   L2  L3
+
+L0 := sha256(concat(0x00, R0))
+L1 := sha256(concat(0x00, R1))
+L2 := sha256(concat(0x00, R2))
+L3 := sha256(concat(0x00, R3))
+Nα := sha256(concat(0x01, hash(L0), hash(L1)))
+Nβ := sha256(concat(0x01, hash(L2), hash(L3)))
+Nγ := sha256(concat(0x01, hash(Nα), hash(Nβ)))
+Nδ := sha256(concat(hash(Nγ),len([L0, L1, L2, L3])))
+
+Receipt tree with five receipts as leaf nodes [L0, L1, L2, L3, L4]
+where R0, R1, R2, R3, R4 are the receipts and Nτ is the root.
+                Nτ   
+              /   \
+            /      \
+          Nζ        N(receipts)
+         /  \
+        /    \
+       Nδ     Iε
+      /  \     \\
+     /    \     \\
+   Nα      Nβ    Nγ
+  /  \    /  \   ||
+ L0  L1  L2  L3  L4
+
+L0 := sha256(concat(0x00, R0))
+L1 := sha256(concat(0x00, R1))
+L2 := sha256(concat(0x00, R2))
+L3 := sha256(concat(0x00, R3))
+L4 := sha256(concat(0x00, R4))
+Nα := sha256(concat(0x01, hash(L0), hash(L1)))
+Nβ := sha256(concat(0x01, hash(L2), hash(L3)))
+Nγ := sha256(concat(0x01, hash(L4), hash(L4)))
+Nδ := sha256(concat(0x01, hash(Nα), hash(Nβ)))
+Nζ := sha256(concat(0x01, hash(Nδ), hash(Iε)))
+Nτ := sha256(concat(hash(Nζ),len([L0, L1, L2, L3, L4])))
+
+Here 'Nτ' is the root generated after concatenating
+the node 'Nζ' with the length of vector of leaf nodes (which is five in the above
+illustration) of the tree and hashing it.
+```
+
+[Link to the specification](https://github.com/solana-foundation/specs/blob/main/core/merkle-tree.md)
+
+#### Benchmarks
+
+We have performed benchmarks comparing two merkle tree implementations,
+the benchmark was done on 500k leaves, each leaf consisted of a 32 byte
+message hash and a single byte status.
+
+1) Solana Labs Merkle Tree: This is the pure rust implementation that is currently
+   used by the Solana Labs client.
+   More details [Solana labs repository](https://github.com/solana-labs/solana/tree/master/merkle-tree)
+2) Firedancer binary merkle tree (bmtree): Implemented in C and uses firedancer's
+   optimised SHA-256 implementation as it's hashing algorithm. However the benchmarks
+   were performed using its rust FFI bindings.
+   More details: [Firedancer](https://github.com/firedancer-io/firedancer/tree/main/src/ballet/bmtree)
+
+![TINY_bench](https://github.com/tinydancer-io/solana-improvement-documents/assets/50767810/637dc83a-b3d2-4616-b70e-4fbb8a9e17fd)
+
+More details with an attached flamegraph can be found in our [repository](https://github.com/tinydancer-io/merkle-bench).
+
+## Impact
+
+This would enable SIMD-0052 to be implemented where we add the receipt root
+to the bankhash. This would allow users to verify their transaction validity
+and status without trusting the RPC.
+
+## Security Considerations
+
+We prepend 0x0 to leaf nodes and 0x1 to internal nodes to avoid second
+preimage attacks where a proof is provided with internal nodes as leaf nodes.
+
+We also make sure that the proof doesn't contain any consecutive identical hashes.
```

Comment:
"proof" in both of these instances in an inclusion proof? let's be explicit again